File: packages/sdk-typescript/src/core/idempotency.ts
import { generateRequestId } from './request-id';

export interface ResolveIdempotencyArgs {
  supplied: string | undefined;
  maxRetries: number;
}

export function resolveIdempotencyKey({
  supplied,
  maxRetries,
}: ResolveIdempotencyArgs): string | null {
  if (supplied) return supplied;
  if (maxRetries > 0) return generateRequestId();
  return null;
}

File: packages/sdk-python/src/bimpeai/_idempotency.py
from __future__ import annotations

from ._request_id import generate_request_id


def resolve_idempotency_key(supplied: str | None, max_retries: int) -> str | None:
    if supplied:
        return supplied
    if max_retries > 0:
        return generate_request_id()
    return None
